File: largest-bst-subtree/largest-bst-subtree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
       
public class Solution {
    public int LargestBSTSubtree(TreeNode root) {
        return Helper(root)[2];
    }
    private int[] Helper(TreeNode node) {
        if(node == null) {
            return new int[]{Int32.MaxValue, Int32.MinValue, 0};
        }
        int[] left = Helper(node.left);
        int[] right = Helper(node.right);
        if(node.val > left[1] && node.val < right[0]) {
            return new int[]{Math.Min(left[0], node.val), Math.Max(node.val, right[1]), left[2] + right[2] + 1};
        }
        else {
            return new int[]{Int32.MinValue, Int32.MaxValue, Math.Max(left[2], right[2])};
        }
    }
}

File: largest-bst-subtree/largest-bst-subtree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int largestBSTSubtree(TreeNode root) {
        int[] res = dfs(root);//count, min, max
        return res[0];
    }
    private int[] dfs(TreeNode node) {
        if(node == null) {
            return new int[]{0, Integer.MAX_VALUE, Integer.MIN_VALUE};
        }
        int[] left = dfs(node.left);
        int[] right = dfs(node.right);
        if(node.val > left[2] && node.val < right[1]) {
            return new int[]{left[0] + right[0] + 1, Math.min(node.val, left[1]), Math.max(node.val, right[2])};
        }
        return new int[]{Math.max(left[0], right[0]), Integer.MIN_VALUE, Integer.MAX_VALUE};
    }
}